### DISCUSSION TITLE
PyTorch Engine has dll file loading error from publish.djl.ai web server on Windows

### DISCUSSION THREAD

**Discussion by iamuspace** [deepjavalibrary/djl#2413] [Development]
My env : 
Windows 10, AMD cpu 64bit, Zulu 64bit JVM, Nvidia Titan X GPU, Cuda 11.7 version, maven 

pytorch engine can not download djl_torch.dll file of publish.djl.ai web server.

Exception in thread "main" ai.djl.engine.EngineException: Cannot download jni files: https://publish.djl.ai/pytorch/1.13.1/jnilib/0.20.0/win-x86_64/cu117/djl_torch.dll
	at ai.djl.pytorch.jni.LibUtils.downloadJniLib(LibUtils.java:515)
	at ai.djl.pytorch.jni.LibUtils.findJniLibrary(LibUtils.java:252)
	at ai.djl.pytorch.jni.LibUtils.loadLibrary(LibUtils.java:80)
	at ai.djl.pytorch.engine.PtEngine.newInstance(PtEngine.java:54)
	at ai.djl.pytorch.engine.PtEngineProvider.getEngine(PtEngineProvider.java:40)
	at test.djl.pytorch.Test.main(Test.java:16)
Caused by: java.io.FileNotFoundException: https://publish.djl.ai/pytorch/1.13.1/jnilib/0.20.0/win-x86_64/cu117/djl_torch.dll
	at java.base/sun.net.www.protocol.http.HttpURLConnection.getInputStream0(HttpURLConnection.java:1989)
	at java.base/sun.net.www.protocol.http.HttpURLConnection.getInputStream(HttpURLConnection.java:1585)
	at java.base/sun.net.www.protocol.https.HttpsURLConnectionImpl.getInputStream(HttpsURLConnectionImpl.java:224)
	at java.base/java.net.URL.openStream(URL.java:1162)
	at ai.djl.util.Utils.openUrl(Utils.java:459)
	at ai.djl.util.Utils.openUrl(Utils.java:443)
	at ai.djl.pytorch.jni.LibUtils.downloadJniLib(LibUtils.java:509) 

--- pom.xml --- 

    <dependencyManagement>
    <dependencies>
        <dependency>
            <groupId>ai.djl</groupId>
            <artifactId>bom</artifactId>
            <version>0.21.0</version>
            <type>pom</type>
            <scope>import</scope>
        </dependency>
    </dependencies>
</dependencyManagement>
    <dependencies>
        <dependency>
            <groupId>ai.djl</groupId>
            <artifactId>api</artifactId>
            <version>0.20.0</version>
        </dependency>
        <dependency>
            <groupId>ai.djl.pytorch</groupId>
            <artifactId>pytorch-engine</artifactId>
            <version>0.20.0</version>
        </dependency>
        <dependency>
            <groupId>ai.djl</groupId>
            <artifactId>basicdataset</artifactId>
            <version>0.20.0</version>
        </dependency>
        <dependency>
            <groupId>ai.djl.pytorch</groupId>
            <artifactId>pytorch-native-cu117</artifactId>
            <version>1.13.1</version>
            <classifier>win-x86_64</classifier>
        </dependency>
    </dependencies>

--- web access by my web browser -----
URL :  https://publish.djl.ai/pytorch/1.13.1/jnilib/0.20.0/win-x86_64/cu117/djl_torch.dll  

<Error>
<Code>NoSuchKey</Code>
<Message>The specified key does not exist.</Message>
<Key>publish/pytorch/1.13.1/jnilib/0.20.0/win-x86_64/cu117/djl_torch.dll</Key>
<RequestId>JN9BSYKQ99J20WB3</RequestId>
<HostId>vZDOP/Hd2jR6xdjbU4smGCJZ5Sf19OI2ykONUc+U7HW6VHXGWdXUzXiBUIwPPy54/fhyfyC41IFFDj9RAdvtkw==</HostId>
</Error>

--- code ---

public class Test {
	public static void main(String[] args) {		
		Engine engine = new PtEngineProvider().getEngine(); 
	}	
}

**Comment by iamuspace** [comment]
pom.xml 


    <dependencyManagement>
    <dependencies>
        <dependency>
            <groupId>ai.djl</groupId>
            <artifactId>bom</artifactId>
            <version>0.21.0</version>
            <type>pom</type>
            <scope>import</scope>
        </dependency>
    </dependencies>
</dependencyManagement>
    <dependencies>
        <dependency>
            <groupId>ai.djl</groupId>
            <artifactId>api</artifactId>
            <version>0.20.0</version>
        </dependency>
        <dependency>
            <groupId>ai.djl.pytorch</groupId>
            <artifactId>pytorch-engine</artifactId>
            <version>0.20.0</version>
        </dependency>
        <dependency>
            <groupId>ai.djl</groupId>
            <artifactId>basicdataset</artifactId>
            <version>0.20.0</version>
        </dependency>
        <dependency>
            <groupId>ai.djl.pytorch</groupId>
            <artifactId>pytorch-native-cu117</artifactId>
            <version>1.13.1</version>
            <classifier>win-x86_64</classifier>
        </dependency>
    </dependencies>

**Reply by iamuspace** [reply]
```
    <dependencyManagement>
    <dependencies>
        <dependency>
            <groupId>ai.djl</groupId>
            <artifactId>bom</artifactId>
            <version>0.21.0</version>
            <type>pom</type>
            <scope>import</scope>
        </dependency>
    </dependencies>
</dependencyManagement>
    <dependencies>
        <dependency>
            <groupId>ai.djl</groupId>
            <artifactId>api</artifactId>
            <version>0.20.0</version>
        </dependency>
        <dependency>
            <groupId>ai.djl.pytorch</groupId>
            <artifactId>pytorch-engine</artifactId>
            <version>0.20.0</version>
        </dependency>
        <dependency>
            <groupId>ai.djl</groupId>
            <artifactId>basicdataset</artifactId>
            <version>0.20.0</version>
        </dependency>
        <dependency>
            <groupId>ai.djl.pytorch</groupId>
            <artifactId>pytorch-native-cu117</artifactId>
            <version>1.13.1</version>
            <classifier>win-x86_64</classifier>
        </dependency>
    </dependencies>
```

**Comment by frankfliu** [comment]
@iamuspace 
1. when you using BOM, you should not specify versions in each dependency, you should let BOM to manage version for you
2. You can not mix 0.20.0 and 0.21.0
3. DJL 0.20.0 doesn't support PyTorch 1.13.1, it only support 1.13.0, you have to use 0.21.0 with PyTorch 1.13.1, see: https://docs.djl.ai/master/engines/pytorch/pytorch-engine/index.html#installation
4. pytorch-native-cu117 is not required, you can remove it in your dependency

**Reply by iamuspace** [reply]
WOW !! runs fine !!! 
Thank you so much frankfliu again. 

key point is BOM manager and delete version of dependences ! 
and delete native-uauto and native-cu. 

solution is so simple, but hard to know this before. 

Thanks frankfliu !!! Good Luck !
